be/src/storage/tablet_updates.cpp

```diff
@@ -2620,6 +2623,209 @@ Status TabletUpdates::compaction(MemTracker* mem_tracker) {
+                e.num_rows = stat.num_rows;
+                e.num_dels = stat.num_dels;
+                e.bytes = stat.byte_size;
+                e.num_segments = stat.num_segments;
+            }
+        }
+    }
+
+    struct PKSizeTieredLevel {
+        PKSizeTieredLevel(std::unique_ptr<CompactionInfo> f, int64_t s, int64_t l, int64_t ts, int64_t r, int64_t b,
+                          int64_t rc, int64_t bc)
+                : info(std::move(f)),
+                  score(s),
+                  level(l),
+                  total_segments(ts),
+                  total_rows(r),
+                  total_bytes(b),
+                  total_rows_after_compaction(rc),
+                  total_bytes_after_compaction(bc) {}
+        std::unique_ptr<CompactionInfo> info;
+        int64_t score;
+        int64_t level;
+        int64_t total_segments;
+        int64_t total_rows;
+        int64_t total_bytes;
+        int64_t total_rows_after_compaction;
+        int64_t total_bytes_after_compaction;
+    };
+
+    struct LevelComparator {
+        bool operator()(const PKSizeTieredLevel* left, const PKSizeTieredLevel* right) const {
+            return left->score > right->score || (left->score == right->score && left->level < right->level);
+        }
+    };
+    std::vector<std::unique_ptr<PKSizeTieredLevel>> order_levels;
+    std::set<PKSizeTieredLevel*, LevelComparator> priority_levels;
+
+    for (auto& [level, candidates] : candidates_by_level) {
+        std::unique_ptr<CompactionInfo> level_info = std::make_unique<CompactionInfo>();
+        level_info->start_version = version;
+        std::sort(candidates.begin(), candidates.end());
+
+        int64_t score = 0;
+        int64_t total_segments = 0;
+        int64_t total_rows = 0;
+        int64_t total_bytes = 0;
+        int64_t total_rows_after_compaction = 0;
+        int64_t total_bytes_after_compaction = 0;
+        for (auto& e : candidates) {
+            size_t new_rows = total_rows_after_compaction + e.num_rows - e.num_dels;
+            size_t new_bytes = total_bytes_after_compaction;
+            if (e.num_rows != 0) {
+                new_bytes += e.bytes * (e.num_rows - e.num_dels) / e.num_rows;
+            }
+            if (total_bytes_after_compaction > 0 && new_bytes > config::update_compaction_result_bytes * 2) {
+                break;
+            }
+            level_info->inputs.push_back(e.rowsetid);
+            score += e.score_per_row * (e.num_rows - e.num_dels);
+            total_rows_after_compaction = new_rows;
+            total_bytes_after_compaction = new_bytes;
+            total_rows += e.num_rows;
+            total_bytes += e.bytes;
+            total_segments += e.num_segments;
+            if (total_bytes_after_compaction > config::update_compaction_result_bytes ||
+                level_info->inputs.size() >= config::max_update_compaction_num_singleton_deltas) {
+                break;
+            }
```

Comment:
better use all rowsets in this level to new `PKSizeTieredLevel` and calculate score,
then select some (or all) rowsets from the max level by these checks.

Reply:
> better use all rowsets in this level to new `PKSizeTieredLevel` and calculate score, then select some (or all) rowsets from the max level by these checks.

done.